src/Blocks/sources.jl

```diff
@@ -730,3 +731,145 @@ end
+        cache.bufs[(u, x)] = interpolation_type(
+            get_tmp(prev_u, u), get_tmp(prev_x, x), args...)
+    else
+        # @info "cache hit"
+        cache[(u, x)]
+    end
+
+    return interp
+end
+
+# function Symbolics.derivative(::typeof(CachedInterpolation), args::NTuple{2, Any}, ::Val{2})
+#     Symbolics.derivative(args[1], (args[2],), Val(1))
+# end
+
+Base.nameof(::CachedInterpolation) = :CachedInterpolation
+
+@register_symbolic (f::CachedInterpolation)(u::AbstractArray, x::AbstractArray, args::Tuple)
+
+"""
+    ParametrizedInterpolation(interp_type, u, x, args...; name, t = ModelingToolkit.t_nounits)
+
+Represent function interpolation symbolically as a block component, with the interpolation data represented parametrically.
+By default interpolation types from [`DataInterpolations.jl`](https://github.com/SciML/DataInterpolations.jl) are supported,
+but in general any callable type that builds the interpolation object via `itp = interpolation_type(u, x, args...)` and calls
+the interpolation with `itp(t)` should work. This does not need to represent an interpolation, it can be any type that satisfies
+the interface, such as lookup tables.
+# Arguments:
+  - `interp_type`: the type of the interpolation. For `DataInterpolations`,
+these would be any of [the available interpolations](https://github.com/SciML/DataInterpolations.jl?tab=readme-ov-file#available-interpolations),
+such as `LinearInterpolation`, `ConstantInterpolation` or `CubicSpline`.
+  - `u`: the data used for interpolation. For `DataInterpolations` this will be an `AbstractVector`
+  - `x`: the values that each data points correspond to, usually the times corresponding to each value in `u`.
+  - `args`: any other arguments beeded to build the interpolation
+# Keyword arguments:
+  - `name`: the name of the component
+
+# Parameters:
+  - `data`: the symbolic representation of the data passed at construction time via `u`.
+  - `ts`: the symbolic representation of times corresponding to the data passed at construction time via `x`.
+
+# Connectors:
+  - `input`: a [`RealInput`](@ref) connector corresponding to the independent variable
+  - `output`: a [`RealOutput`](@ref) connector corresponding to the interpolated value
+"""
+function ParametrizedInterpolation(
+        interp_type::T, u::AbstractVector, x::AbstractVector, args...;
+        name) where {T}
+    build_interpolation = CachedInterpolation(interp_type, u, x, args)
+
+    @parameters data[1:length(x)] = u
+    @parameters ts[1:length(x)] = x
+    @parameters interpolation_type::T=interp_type [tunable = false]
+    @parameters (interpolator::interp_type)(..)::eltype(u)
+
+    @named input = RealInput()
+    @named output = RealOutput()
+
+    eqs = [output.u ~ interpolator(input.u)]
+
+    ODESystem(eqs, ModelingToolkit.t_nounits, [],
+        [data, ts, interpolation_type, interpolator];
+        parameter_dependencies = [
+            interpolator ~ build_interpolation(data, ts, args)
```

Comment:
Interpolating a struct into a symbolic expression will result in that struct being constructed in the codegen. i.e. this will codegen as something like
```julia
(CachedInterpolation{...}(interp_type, [...], [...], args))(data, ts, args)
```
Which will result in the arrays in `[...]` being allocated every f-call, and the constructor being called every `f-call`. This can be avoided by making `build_interpolation` a callable parameter.

Reply:
More concretely, I added some `@info`s in the `for (sym, val) in p` loop
```julia
@info "sym: $sym"
sym = unwrap(sym)
@info "val: $val"
val = unwrap(val)
ctype = symtype(sym)
@info "ctype: $ctype"
if symbolic_type(val) !== NotSymbolic()
    error("Could not evaluate value of parameter $sym. Missing values for variables in expression $val.")
end
if ctype <: FnType
    ctype = fntype_to_function_type(ctype)
end
@info "final ctype: $ctype"
val = symconvert(ctype, val)
```
and the result is
```
[ Info: sym: i₊interpolator⋆
[ Info: val: ModelingToolkitStandardLibrary.Blocks.CachedInterpolation{LinearInterpolation{Vector{Float64}, Vector{Float64}, Vector{Float64}, Vector{Float64}, Float64, (1,)}, UnionAll, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, PreallocationTools.GeneralLazyBufferCache{ModelingToolkitStandardLibrary.Blocks.var"#123#124"{UnionAll, Tuple{}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}}}}(LinearInterpolation, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}([0.32444839943667103, 0.20600750899157028, 0.7361716345389965, 0.8547116978444484, 0.7436176823347252, 0.8924785257096608, 0.8080492576860319, 0.8717347592230256, 0.161408199841236, 0.31607678907385683, 0.7914505075610498, 0.7860058028324105, 0.15325190632974606, 0.11313110513482694, 0.7237555821387326], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], Any[]), PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}([0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], Any[]), PreallocationTools.GeneralLazyBufferCache{ModelingToolkitStandardLibrary.Blocks.var"#123#124"{UnionAll, Tuple{}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}}}(Dict{Any, Any}(Tuple{Vector{Float64}, Vector{Float64}} => LinearInterpolation{Vector{Float64}, Vector{Float64}, Vector{Float64}, Vector{Float64}, Float64, (1,)}([0.32444839943667103, 0.20600750899157028, 0.7361716345389965, 0.8547116978444484, 0.7436176823347252, 0.8924785257096608, 0.8080492576860319, 0.8717347592230256, 0.161408199841236, 0.31607678907385683, 0.7914505075610498, 0.7860058028324105, 0.15325190632974606, 0.11313110513482694, 0.7237555821387326], [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0], Float64[], DataInterpolations.LinearParameterCache{Vector{Float64}}(Float64[]), false, FindFirstFunctions.Guesser{Vector{Float64}}([0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0], Base.RefValue{Int64}(1), true), false, false)), ModelingToolkitStandardLibrary.Blocks.var"#123#124"{UnionAll, Tuple{}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}}(LinearInterpolation, (), PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}([0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], Any[]), PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}([0.32444839943667103, 0.20600750899157028, 0.7361716345389965, 0.8547116978444484, 0.7436176823347252, 0.8924785257096608, 0.8080492576860319, 0.8717347592230256, 0.161408199841236, 0.31607678907385683, 0.7914505075610498, 0.7860058028324105, 0.15325190632974606, 0.11313110513482694, 0.7237555821387326], [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], Any[]))))
[ Info: ctype: SymbolicUtils.FnType{Tuple, Float64, LinearInterpolation}
[ Info: final ctype: LinearInterpolation
ERROR: MethodError: Cannot `convert` an object of type 
  ModelingToolkitStandardLibrary.Blocks.CachedInterpolation{LinearInterpolation{Vector{Float64}, Vector{Float64}, Vector{Float64}, Vector{Float64}, Float64, (1,)}, UnionAll, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, PreallocationTools.GeneralLazyBufferCache{ModelingToolkitStandardLibrary.Blocks.var"#123#124"{UnionAll, Tuple{}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}, PreallocationTools.DiffCache{Vector{Float64}, Vector{Float64}}}}} to an object of type 
  LinearInterpolation
The function `convert` exists, but no method is defined for this combination of argument types.
```

Reply:
This performance issue is expected and is documented. I don't think we have a better solution right now for the case of differentiating w.r.t. parameters, so I think it's fine.